src/composer.ts

```diff
@@ -12,6 +13,34 @@ type Triggers<C> = MaybeArray<
 type Predicate<T> = (t: T) => boolean
 type AsyncPredicate<T> = (t: T) => Promise<boolean>
 
+type UnionToIntersection<U> = (
+  U extends unknown ? (k: U) => void : never
+) extends (k: infer I) => void
+  ? I
+  : never
+type PropOr<T, P extends string, D> = T extends Record<P, infer V> ? V : D
+type GuardedContext<
+  C extends Context,
+  U extends Omit<tt.Update, 'update_id'>
+> = C &
+  {
+    [P in tt.UpdateType as SnakeToCamelCase<P>]: PropOr<U, P, undefined>
+  } & {
+    update: U
+    updateType: keyof UnionToIntersection<U>
+  }
+type MountMap = {
```

Comment:
I would like to find a better name for this, or add an explanation that highlights its usage so the name becomes clearer. The type basically maps `T in tt.UpdateType | tt.MesssageSubType` to the correct properties on an `Update` that need to be intersected with `Context` in order to narrow it down according to `T`, correct? This makes it correspond to the superseded `GuaranteedContextProps` type, so that would be my first suggestion.

Alternatively, we could add a comment that explains its usage with a terminology that points out what mounting and mapping would mean here.

Reply:
> correct?

Yes.

Name is good, it's a mapping used exclusively by `Composer.mount` and `Composer::on`, I'll add some JSDoc tho.

Reply:
Isn't it a part of `MatchedContext`? So it is basically becoming the central point of type inference for all `Composer` methods, together with `GuardedContext`. It's hence not exclusively used by those two methods